py_mypy/README.md

```diff
@@ -0,0 +1,109 @@
+
+Notionally how this will all work is that:
+
+- We need to create an aspect configured to use whatever `mypy` tool we may want
+- That aspect will extend the `py_*` rules in the build graph to add an output group capturing the MyPy typecheck cache.
+  These typecheck cache outputs will depend on the typecheck cache outputs of all dependencies.
+- We will configure our `.bazelrc` to apply the aspect so that users don't have to think about it.
+- We will configure our `.bazelrc` to select the typecheck output group in addition to the default outputs of rules.
+
+This all has the effect of creating a build sub-graph parallel to our normal build graph which instead of producing and consuming Python files as dependencies produces and consumes the MyPy analysis caches.
+
+To take a simple example, let's say that we have a small build graph
+
+```mermaid
+graph TD
+    A[data_models] --> B[data_persistence];
+    A --> C[inventory_management];
+    A --> D[order_processing];
+    B --> C;
+    B --> D;
+    C --> E[cli];
+    D --> E;
+    F["click (3rdparty)"] --> E;
+    G["pydantic (3rdparty)"] --> A;
+```
+
+Ordinarily the dependencies between these rules take the form of the Python source files underlying the rules.
+But when we activate our `mypy` aspect and select the `mypy` output group, the cache files from typechecking each* of these targets also become part of that dependency chain.
+This allows Bazel to drive typechecking these libraries in depgraph order while caching intermediate results.
+
+We can demonstrate this by looking at the results of `bazel aquery`, which will show MyPy invocations and that the resulting cache trees are dependencies between each of the invocations.
+But more on that in a minute.
+
+## Setup
+
+In this example we've set up `rules_python` in combination with `rules_uv`, which provides lockfile compilation.
+These two give us a Python dependency solution (including the MyPy we want to use), which we'll feed into `rules_mypy`.
+
+The main trick is in `//tools/mypy:BUILD.bazel`, where we provide a definition of the MyPy CLI binary which we can feed into the checking aspect.
+This is important because it allows us to use our locked requirement for MyPy, and to provide MyPy plugins.
+If we didn't do this, `rules_mypy` would "helpfully" provide an embedded default version and configuration of MyPy which may or may not be what we want.
+
+We also need the `.bazelrc` as previously discussed to enable both the aspect and the typecheck output group.
+
+Since there's other Python code in this monorepo which doesn't typecheck and we don't want to have to address that to adopt typing, we're going to use the `opt_in_tags` parameter on the aspect configuration.
+This allows us to specify `tags=["mypy"]` on relevant Python targets to selectively apply typechecking rather than just getting mypy checks applied to everything.
```

Comment:
What's the alternative? is there an `opt_out_tags` instead? I'm pretty sure we want that, since the most important bit is "what about new code" - most teams want new code to be typed by default to prevent back-sliding